### PR TITLE
[CARBONDATA-3684] Remove MDK and cardinality in write path

### PR DESCRIPTION
**Pull request by jackylk**
### Why is this PR needed?
 Since only DATE is dictionary now, MDK is always 4 bytes. We can simplify many places in write path to remove MDK and cardinality calculation to save memory
 
 ### What changes were proposed in this PR?
1. MDK generation is removed
2. cardinality calculation is removed
3. some other unused declaration is removed
    
 ### Does this PR introduce any user interface change?
 - No

 ### Is any new testcase added?
 - No

### CODE REVIEW

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1808/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1812/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1813/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1814/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/1/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/145/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/146/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/147/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1851/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/148/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1852/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/159/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1862/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/160/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1863/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/161/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1864/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/174/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1877/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/175/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1878/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/183/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1886/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/186/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1889/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/189/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1891/

File: integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/CarbonMergerRDD.scala [hunk 1]
```diff
@@ -578,15 +562,11 @@ class CarbonMergerRDD[K, V](
       }
     }
     val updatedMaxSegmentColumnList = new util.ArrayList[ColumnSchema]()
-    // update cardinality and column schema list according to master schema
-    val cardinality = CarbonCompactionUtil
```
Comment (QiangCai) [inline integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/CarbonMergerRDD.scala]: if no need to update cardinality,  also no need to update column schema

File: integration/spark-common/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchemaCommon.scala [hunk 2]
```diff
@@ -111,8 +111,6 @@ case class CarbonMergerMapping(
     validSegments: Array[Segment],
     tableId: String,
     campactionType: CompactionType,
-    // maxSegmentColCardinality is Cardinality of last segment of compaction
-    var maxSegmentColCardinality: Array[Int],
     // maxSegmentColumnSchemaList is list of column schema of last segment of compaction
     var maxSegmentColumnSchemaList: List[ColumnSchema],
```
Comment (QiangCai) [inline integration/spark-common/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchemaCommon.scala]: use CarbonTable directly

File: core/src/main/java/org/apache/carbondata/core/datamap/DataMapChooser.java [hunk 3]
```diff
@@ -135,25 +135,6 @@ DataMapExprWrapper chooseDataMap(DataMapLevel level, FilterResolverIntf resolver
     return null;
   }
 
-  /**
-   * Get all datamaps of the table for clearing purpose
-   */
-  public DataMapExprWrapper getAllDataMapsForClear(CarbonTable carbonTable)
```
[the remark this answers is not in the document]
Comment (jackylk) [reply core/src/main/java/org/apache/carbondata/core/datamap/DataMapChooser.java]: It is not used anyplace

File: core/src/main/java/org/apache/carbondata/core/datastore/block/SegmentProperties.java [hunk 4]
```diff
@@ -640,15 +377,91 @@ public int getNumberOfSortColumns() {
...
 
-  public int getNumberOfDictSortColumns() {
-    return this.numberOfSortColumns - this.numberOfNoDictSortColumns;
+  public int getNumberOfDictDimensions() {
+    return numberOfDictDimensions;
   }
 
-  public int getLastDimensionColOrdinal() {
-    return lastDimensionColOrdinal;
+  public int getNumberOfSimpleDimensions() {
```
[the remark this answers is not in the document]
Comment (jackylk) [reply core/src/main/java/org/apache/carbondata/core/datastore/block/SegmentProperties.java]: fixed

File: core/src/main/java/org/apache/carbondata/core/datastore/block/TableBlockInfo.java [hunk 5]
```diff
@@ -74,31 +72,10 @@
    */
   private Segment segment;
 
-  /**
-   * id of the Blocklet.
-   */
-  private String blockletId;
```
[the remark this answers is not in the document]
Comment (jackylk) [reply core/src/main/java/org/apache/carbondata/core/datastore/block/TableBlockInfo.java]: It is not used anywhere

File: core/src/main/java/org/apache/carbondata/core/datastore/blocklet/EncodedBlocklet.java [hunk 6]
```diff
@@ -38,11 +37,6 @@
    */
   private int blockletSize;
 
-  /**
-   * list of page metadata
-   */
-  private List<TablePageKey> pageMetadataList;
```
[the remark this answers is not in the document]
Comment (jackylk) [reply core/src/main/java/org/apache/carbondata/core/datastore/blocklet/EncodedBlocklet.java]: yes, it is not used in query flow

File: core/src/main/java/org/apache/carbondata/core/scan/processor/DataBlockIterator.java [hunk 7]
```diff
@@ -217,7 +217,8 @@ public BlockletScannedResult call() throws Exception {
             nextRead.set(true);
             futureIo = readNextBlockletAsync();
           }
-          return blockletScanner.scanBlocklet(rawBlockletColumnChunks);
+          BlockletScannedResult result = blockletScanner.scanBlocklet(rawBlockletColumnChunks);
```
[the remark this answers is not in the document]
Comment (jackylk) [reply core/src/main/java/org/apache/carbondata/core/scan/processor/DataBlockIterator.java]: fixed

File: core/src/main/java/org/apache/carbondata/core/util/ByteUtil.java [hunk 8]
```diff
@@ -756,4 +756,45 @@ public static long toLongLittleEndian(byte[] bytes, int offset) {
...
+
+  public static byte[] convertDateToBytes(long[] date) {
+    byte[] output = new byte[date.length * 4];
+    for (int i = 0; i < date.length; i++) {
+      System.arraycopy(ByteUtil.toBytes(date[i]), 0, output, i * 4, 4);
+    }
+    return output;
+  }
+
+  public static int convertBytesToDate(byte[] date) {
```
[the remark this answers is not in the document]
Comment (jackylk) [reply core/src/main/java/org/apache/carbondata/core/util/ByteUtil.java]: fixed

File: core/src/main/java/org/apache/carbondata/core/util/ByteUtil.java [hunk 9]
```diff
@@ -756,4 +756,45 @@ public static long toLongLittleEndian(byte[] bytes, int offset) {
...
+
+  public static int convertBytesToDate(byte[] date, int offset) {
+    return ByteUtil.toInt(date, offset);
+  }
+
+  public static int dateBytesSize() {
+    return 4;
+  }
+
+  public static int[] convertBytesToDateIntArray(byte[] input) {
```
[the remark this answers is not in the document]
Comment (jackylk) [reply core/src/main/java/org/apache/carbondata/core/util/ByteUtil.java]: fixed

[the remark this answers is not in the document]
Comment (jackylk) [reply core/src/main/java/org/apache/carbondata/core/util/ByteUtil.java]: fixed

File: integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/CarbonMergerRDD.scala (hunk 1 above)
Comment (jackylk) [reply integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/CarbonMergerRDD.scala]: It is needed, otherwise testcase failed

Comment (jackylk) [reply integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/CarbonMergerRDD.scala]: I will remove it and try the CI

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/217/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1919/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/225/

File: core/src/main/java/org/apache/carbondata/core/datastore/block/SegmentProperties.java [hunk 10]
```diff
@@ -640,15 +377,91 @@ public int getNumberOfSortColumns() {
...
+      for (int i = 0; i < depth; i++) {
+        length[index++] = -1;
+      }
+    }
+    for (CarbonMeasure measure : measures) {
+      DataType dataType = measure.getDataType();
+      if (DataTypes.isDecimal(dataType)) {
+        length[index++] = -1;
+      } else {
+        length[index++] = 8;
```
[the remark this answers is not in the document]
Comment (jackylk) [reply core/src/main/java/org/apache/carbondata/core/datastore/block/SegmentProperties.java]: I have created JIRA for this: CARBONDATA-3690, will fix later

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1927/

**Comment by jackylk** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/239/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1941/

**Comment by jackylk** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/242/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1944/

**Comment by QiangCai** [comment]
LGTM

**Pull request closed without merging**